Gems/Atom/RPI/Code/Source/RPI.Public/Material/Material.cpp

```diff
@@ -257,24 +255,12 @@ namespace AZ
             if (newMaterialAsset)
             {
                 Init(*newMaterialAsset);
-                MaterialReloadNotificationBus::Event(newMaterialAsset.GetId(), &MaterialReloadNotifications::OnMaterialReinitialized, this);
```

Comment:
Hm, there could still be a need for a reinitialization bus pattern, at least the OnMaterialReinitialized function. It all depends on whether the AssetBus callback order is the same as the order in which BusConnect is called. If that's _not_ the case, here is how it could fail:

MyComponent wants to know when the Material has reinitialized so the component can reinitialize itself with updated material data. Both Material and MyComponent are connected to the AssetBus. The MaterialTypeAsset is reloaded and calls OnDependencyAssetReloaded. It happens to call MyComponent first, which updates itself using _old_ data from the Material object. Next the Material callback is called and reinitailizes itself. But MyComponent is stuck using the old data.

If the AssetBus::Handler::OnConnect order determines the callback order, then we just have to make sure that Material connects to AssetBus before MyComponent does.

Reply:
BTW, developers will have to be careful with bus disconnect and reconnect. If an asset reload handler were to disconnect and then reconnect to the AssetBus for some reason, that would change the callback order. The Material class should be safe because it doesn't disconnect, it just calls BusConnect again with the same ID which is a no-op. Still, it's a bit nuanced, it might be worth having OnMaterialReinitialized just to be safe and explicit. I'm on the fence about it.

Reply:
The ebus is going to be working with assets, so I'm a little confused on how it maps to your example.  I assume Material depends on MaterialType, but what about MyComponent?  If there's a MyComponentAsset that depends on Material, then the tree looks like MyComponetAsset -> MaterialAsset -> MaterialTypeAsset.  In this case, the order of connection doesn't matter.  MaterialAsset will always be called first as it's the first parent, and MyComponentAsset will always be called second.  If you instead have MyComponentAsset depend on MaterialTypeAsset, then they're at the same level, and I don't think we'll have any stable order there.  But if that's the case, MyComponentAsset should not care about MaterialTypeAsset.  If it does, it needs to set up its dependencies that way.

Reply:
MyComponent is a hypothetical O3DE entity component (you could consider the actual MaterialComponent, but I was just keeping my example more abstract). So there is no "MyComponentAsset". In my example MyComponent has a MaterialAsset, calls Material::Create(m_materialAsset) to create the Material instance. The Material self-initializes using the MaterialAsset. This Material it what's actually used for rendering. Material::OnDependentAssetReloaded will cause the Material to reinitailize itself. Suppose that MyComponent also wants to make some updates after that, using the reinitialized Material object. We have to make sure that MyComponent is notified about the changes _after_ Material::OnDependentAssetReloaded has done it's business.